github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+	data, err := json.Marshal(content)
+	if err != nil {
+		return nil, err
+	}
+
+	respWriter := httptest.NewRecorder()
+	respWriter.WriteHeader(statusCode)
+	_, err = respWriter.Write(data)
+	if err != nil {
+		return nil, err
+	}
+
+	response := respWriter.Result()
+
+	if headers != nil {
+		for k, v := range headers(time.Now()) {
+			response.Header.Set(k, v)
+		}
+	}
+
+	return response, err
+}
+
+func newRequest(url string) *http.Request {
+	req, _ := http.NewRequest("GET", url, http.NoBody)
+	return req
+}
+
+// gitHubTransportResponseMock avoids calling the network to get the response directly from testcases
+type gitHubTransportResponseMock struct{}
+
+func (gh *gitHubTransportResponseMock) RoundTrip(req *http.Request) (*http.Response, error) {
+	url := req.URL.String()
+	tc, ok := testCases[url]
+	if !ok {
+		return newResponse(apiResponse{}, nil, http.StatusOK)
+	}
+
+	if tc.err != nil {
+		return &http.Response{}, tc.err
+	}
+
+	return newResponse(tc.response, tc.headers, tc.statusCode)
+}
+
+func TestRateLimitSuite(t *testing.T) {
+	suite.Run(t, new(RateLimitSuite))
+}
+
+type RateLimitSuite struct {
+	suite.Suite
+	transport  *RateLimitTransport
+	loggerMock *testutils.LoggerMock
+}
+
+func (s *RateLimitSuite) SetupSuite() {
+	defaultAbuseRetryAfter = default403Reset
+}
+
+func (s *RateLimitSuite) SetupTest() {
+	s.loggerMock = &testutils.LoggerMock{}
+	s.transport = NewRateLimitTransport(&gitHubTransportResponseMock{}, s.loggerMock)
+}
+
+func (s *RateLimitSuite) TearDownSuite() {
+}
+
+// TestNoLimit ensures that the regular case: no rate limit nor abuse, is not considered a RateLimit case
+func (s *RateLimitSuite) TestNoLimit() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/normal"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
+
+	content, err := ioutil.ReadAll(response.Body)
+	s.NoError(err)
+	var data apiResponse
+	err = json.Unmarshal(content, &data)
+	s.NoError(err)
+	s.Equal("success", data.Data)
+	s.Len(data.Errors, 0)
+}
+
+// TestRateLimitConsecutively ensures that hitting RateLimit twice, causes two waiting periods
+// it should log the RateLimitError, and wait till the lock expires
+// if the next Request hits the RateLimit again, it should do the same
+// if the next Request does not hit the RateLimit, it should return the response inmediately
+func (s *RateLimitSuite) TestRateLimitConsecutively() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
```

Comment:
could you please explain why the response is empty here?
ratelimit_sleep returns 200 and some response, why do we remove it?

Reply:
We're kind of abusing the Transport chain, processing the responses, and returning errors according to some internal business logic, which is not exactly the purpose of the Transport chain.
Anyhow, taking this into consideration:
- I remember that go discards the response if there's also an error,
- we can consider the `ErrRateLimit`, as nothing to see here.

I'd purpose to merge and open an issue if you prefer reconsidering it.